src/Compilers/Core/Portable/SourceGeneration/SourceGeneratorContext.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+#nullable enable
+namespace Microsoft.CodeAnalysis
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In Progress")]
+    public readonly struct SourceGeneratorContext
+    {
+        internal SourceGeneratorContext(Compilation compilation, AnalyzerOptions options, CancellationToken cancellationToken = default)
+        {
+            Compilation = compilation;
+            AnalyzerOptions = options;
+            CancellationToken = cancellationToken;
+            AdditionalSources = new AdditionalSourcesCollection();
+        }
+
+        public Compilation Compilation { get; }
+
+        // PROTOTYPE: replace AnalyzerOptions with an differently named type that is otherwise identical.
+        // The concern being that something added to one isn't necessarily applicable to the other.
+        public AnalyzerOptions AnalyzerOptions { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public AdditionalSourcesCollection AdditionalSources { get; }
+
+        public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); }
```

Comment:
>public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); } [](start = 8, length = 92)

What is the purpose of this API?

Reply:
We need it for generators to be able to report diagnostics (couldn't generate etc.). We want to get it in the initial shape to show where it'll exist, even though it doesn't actually work yet.

Reply:
We need it for generators to be able to report diagnostics (couldn't generate etc.). We want to get it in the initial shape to show where it'll exist, even though it doesn't actually work yet.

---
In reply to: [382815234](https://github.com/dotnet/roslyn/pull/41789#discussion_r382815234) [](ancestors = 382815234)